src/main/java/org/opensearch/plugin/insights/rules/transport/top_queries/TransportTopQueriesAction.java

```diff
@@ -68,27 +75,143 @@ public TransportTopQueriesAction(
         this.queryInsightsService = queryInsightsService;
     }
 
+    ActionListener<TopQueriesResponse> createInMemoryDataCollectionListener(
+        TopQueriesRequest request,
+        ActionListener<TopQueriesResponse> finalListener
+    ) {
+        return new ActionListener<TopQueriesResponse>() {
+            @Override
+            public void onResponse(TopQueriesResponse inMemoryQueriesResponse) {
+                handleInMemoryDataResponse(request, inMemoryQueriesResponse, finalListener);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                finalListener.onFailure(e);
+            }
+        };
+    }
+
+    void handleInMemoryDataResponse(
+        TopQueriesRequest request,
+        TopQueriesResponse inMemoryQueriesResponse,
+        ActionListener<TopQueriesResponse> finalListener
+    ) {
+        List<TopQueries> inMemoryTopQueries = inMemoryQueriesResponse.getNodes();
+        List<FailedNodeException> inMemoryDataFailures = inMemoryQueriesResponse.failures();
+        String from = request.getFrom();
+        String to = request.getTo();
+        if (from != null && to != null) {
+            fetchHistoricalData(request, inMemoryTopQueries, inMemoryDataFailures, finalListener);
+        } else {
+            finalListener.onResponse(inMemoryQueriesResponse);
+        }
+    }
+
+    void fetchHistoricalData(
+        TopQueriesRequest request,
+        List<TopQueries> inMemoryTopQueries,
+        List<FailedNodeException> inMemoryDataFailures,
+        ActionListener<TopQueriesResponse> finalListener
+    ) {
+        String from = request.getFrom();
+        String to = request.getTo();
+        String id = request.getId();
+        Boolean verbose = request.getVerbose();
+        try (final ThreadContext.StoredContext storedContext = threadPool.getThreadContext().stashContext()) {
+            queryInsightsService.getTopQueriesService(request.getMetricType())
```

Comment:
[Suggestion] I am not sure if exposing `QueryInsightsService`(and all its relevant methods) inside this action is necessary as all it does is to fetch the desired `TopQueriesService`. 
An ideal way would have been to define `TopQueriesServiceFactory` and expose it here, which returns desired service based on metric type. Same could have been used in queryInsightsService.